containers/datalab/content/run.sh

```diff
@@ -63,33 +80,39 @@ setup_tunnel() {
     --ssh-flag="-fNL" \
     --ssh-flag="localhost:8082:localhost:8080" \
     --ssh-key-file="/content/datalab/.config/.ssh/google_compute_engine" \
-    "${ssh_user}@${instance}"
+    "${ssh_user}@${instance}" || kill_container "${ERR_GCLOUD_SSH_FAILED}"
 
   # Test that we can actually call the gateway API via the SSH tunnel
   tunnel_failed=""
   curl -o /tmp/kernel_specs http://localhost:8082/api/kernelspecs 2>/dev/null || tunnel_failed="true"
   if [[ "${tunnel_failed}" == true ]]; then
     echo "Failed to set up the SSH tunnel to the VM ${instance}"
     echo "If the VM was recently created, then it may still be starting up, and retrying the command may work."
-    exit "${ERR_TUNNEL_FAILED}"
+    kill_container "${ERR_TUNNEL_FAILED}"
   fi
 
   default_kernel_spec=`cat /tmp/kernel_specs | python -c $'import json\nprint json.loads(raw_input())["default"]'`
   if [[ -z "${default_kernel_spec}" ]]; then
     echo "Failed to verify that the kernel gateway is running"
     echo "If the VM was recently created, then it may still be starting up, and retrying the command may work."
-    exit "${ERR_GATEWAY_FAILED}"
+    kill_container "${ERR_GATEWAY_FAILED}"
   fi
   echo "Successfully established SSH tunnel to ${instance}"
 
-  watch_and_restart_tunnel &
+  watch_and_restart_tunnel "${project_id}" "${zone}" "${instance}" "${ssh_user}" &
 }
 
 watch_and_restart_tunnel() {
-  while ps --ppid $$ -C ssh 2>&1 > /dev/null; do
+  project_id=$1
+  zone=$2
+  instance=$3
+  ssh_user=$4
+  
+  while [ "$(ps --ppid $$ -o comm= | grep ssh)" != "" ]; do
     sleep 1
```

Comment:
It is probably not a big deal but 5 seconds as interval might save some CPU cycles on the VM (don't know the cost of ps command). Establishing ssh tunnel itself takes more than several seconds anyway.